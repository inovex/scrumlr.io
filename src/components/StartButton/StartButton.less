@import '../../mixins';
@import '../../constants/style';

.start-button {
  display: flex;
  min-height: 50px;

  margin-bottom: @default-margin;
}

.start-button__start {
  flex: 1 0 auto;
  box-sizing: border-box;
  border: none;
  outline: none;
  font-size: 16px;
  font-weight: 700;
  color: @color-white;
  padding: 9px 24px;
  align-self: stretch;

  background-color: @theme-color-primary;
  border-top-left-radius: 3px;
  border-bottom-left-radius: 3px;
  transition: background 0.15s ease;
  cursor: pointer;
  text-transform: uppercase;
}

.start-button__start:focus,
.start-button__start:hover {
  background: @color-dark-mint;
}

.start-button__start-label {
  margin: 0;
  font-size: 1em;
}

.start-button__start-mode {
  font-size: .75em;
}

.start-button__dropdown-toggle {
  box-sizing: border-box;
  border: 2px solid @color-mint;
  outline: none;
  font-size: 16px;
  font-weight: 700;
  color: @color-mint;
  padding: 9px;

  background-color: @color-white;
  border-left: none;
  border-top-right-radius: 3px;
  border-bottom-right-radius: 3px;
  transition: background 0.15s ease;
  cursor: pointer;
  text-transform: uppercase;
}

.start-button__dropdown-toggle-button {
  position: absolute;
  top: -10px;
  left: -16px;
}

.start-button__dropdown-toggle svg {
  width: 32px;
  height: 32px;
}

.start-button__dropdown-toggle svg * {
  fill: currentColor;
  transition: fill 0.15s ease;
}

.start-button__dropdown-toggle:focus,
.start-button__dropdown-toggle:hover {
  background: @color-dark-mint;
  color: @color-white;
}

.start-button__dropdown-wrapper {
  box-sizing: border-box;
  margin-left: 0;
  flex: 0 0 auto;
  align-self: stretch !important;
}