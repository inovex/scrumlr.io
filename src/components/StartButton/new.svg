<svg width="38" height="22" viewBox="0 0 127 75" xmlns="http://www.w3.org/2000/svg" fill-rule="evenodd" clip-rule="evenodd" stroke-linejoin="round" stroke-miterlimit="1.414"><path fill="none" d="M0 0h126.055v74.728H0z"/><path fill="#c00" d="M4.07 23.525l106.685-19.73 8.726 47.184L12.796 70.71z"/><g fill="#fff" fill-rule="nonzero"><path d="M23.99 57.37c-1.65.306-2.546.068-2.692-.717l-4.418-23.89c-.074-.405.07-.747.435-1.025.362-.275.967-.49 1.815-.647l1.756-.325c.663-.122 1.224-.156 1.68-.096.46.06.866.266 1.218.624L36.008 43.08l1.388 1.307-2.757-14.908c-.075-.405.07-.747.435-1.025.362-.275.967-.49 1.815-.647l1.18-.219c.848-.156 1.487-.17 1.918-.043.429.132.682.4.757.805l4.418 23.89c.145.784-.6 1.326-2.233 1.628l-1.57.29c-.68.126-1.24.163-1.683.108-.439-.056-.835-.257-1.183-.597L25.706 41.424l-1.097-1.012 2.795 15.112c.145.784-.6 1.326-2.234 1.628l-1.18.218zM52.697 51.999c-.568.105-1.009.064-1.313-.128-.305-.188-.509-.543-.604-1.06l-4.283-23.155c-.096-.52-.03-.917.188-1.194.217-.281.614-.473 1.182-.578L62.923 23.1c.43-.08.769.01 1.017.267.247.258.43.717.552 1.373l.21 1.136c.12.655.114 1.15-.025 1.479-.139.33-.423.534-.854.614l-10.359 1.915 1.002 5.417 9.25-1.71c.43-.08.77.01 1.017.267.248.258.43.717.552 1.373l.214 1.157c.121.656.115 1.15-.025 1.48-.139.33-.423.534-.853.613l-9.25 1.71 1.118 6.049 10.383-1.92c.43-.08.769.01 1.017.267.247.258.429.71.547 1.351l.218 1.175c.122.66.115 1.154-.024 1.483-.14.33-.424.534-.854.614l-15.08 2.789zM79.346 47.133c-.663.123-1.195.155-1.607.094-.41-.062-.668-.214-.777-.453L66.036 24.02a1.525 1.525 0 0 1-.2-.549c-.056-.308.112-.587.502-.841.394-.254 1.021-.46 1.885-.62l1.062-.195c1.602-.297 2.594-.076 2.968.666l5.299 11.437c.082.203.414.96.99 2.27l1.274 2.869c-.003-1.056.09-2.54.274-4.457l.117-1.174 1.112-13.014c.007-.29.19-.536.546-.731.356-.196.87-.354 1.548-.48l1.575-.29c.675-.125 1.215-.162 1.617-.107.402.055.657.22.771.487l5.55 11.783a377.347 377.347 0 0 1 2.454 5.367c.013-.265.067-1.187.158-2.767l.187-3.117.972-12.597c.084-.827.931-1.387 2.533-1.683l.462-.086c.849-.157 1.5-.188 1.959-.092.454.098.712.299.769.607.018.096.028.195.028.287l.005.292-2.274 25.194c-.033.266-.242.504-.636.714-.39.21-.887.368-1.487.479l-3.007.556c-.6.111-1.12.14-1.563.085-.44-.055-.716-.204-.821-.444l-5.767-12.288-1.116-2.42-.249 2.673-1.1 13.557c-.033.266-.242.504-.636.714-.39.21-.887.368-1.49.48l-2.96.547z"/></g></svg>