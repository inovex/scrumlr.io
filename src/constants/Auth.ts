import * as firebase from 'firebase/app';

export interface AuthProviderMap {
  google: firebase.auth.AuthProvider;
  twitter: firebase.auth.AuthProvider;
  github: firebase.auth.AuthProvider;
}

export type AuthProvider = keyof AuthProviderMap;

export const instantiateAuthProviders = (
  firebase: firebase.app.App
): AuthProviderMap => ({
  google: new firebase.auth['GoogleAuthProvider'](),
  twitter: new firebase.auth['TwitterAuthProvider'](),
  github: new firebase.auth['GithubAuthProvider']()
});
