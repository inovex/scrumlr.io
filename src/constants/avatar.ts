import {
  AVATAR_ACCESSORIES_TYPES,
  AVATAR_CLOTHE_COLORS,
  AVATAR_CLOTHE_TYPES,
  AVATAR_EYE_TYPES,
  AVATAR_EYEBROW_TYPES,
  AVATAR_FACIAL_HAIR_COLORS,
  AVATAR_FACIAL_HAIR_TYPES,
  AVATAR_GRAPHIC_TYPES,
  AVATAR_HAIR_COLORS,
  AVATAR_MOUTH_TYPES,
  AVATAR_SKIN_COLORS,
  AVATAR_TOP_TYPES,
  AvatarConfig,
} from "../components/Avatar/types";

export const AVATAR_CONFIG: AvatarConfig = {
  hair: [
    {values: AVATAR_TOP_TYPES, key: "topType"},
    {
      values: AVATAR_HAIR_COLORS,
      key: "hairColor",
      disabledOn: {topType: ["NoHair", "Eyepatch", "Hat", "Hijab", "Turban", "WinterHat1", "WinterHat2", "WinterHat3", "WinterHat4", "LongHairFrida", "LongHairShavedSides"]},
    },
    {values: AVATAR_FACIAL_HAIR_TYPES, key: "facialHairType", disabledOn: {topType: ["Hijab"]}},
    {values: AVATAR_FACIAL_HAIR_COLORS, key: "facialHairColor", disabledOn: {topType: ["Hijab"], facialHairType: ["Blank"]}},
  ],
  facialFeatures: [
    {values: AVATAR_SKIN_COLORS, key: "skinColor"},
    {values: AVATAR_EYEBROW_TYPES, key: "eyebrowType"},
    {values: AVATAR_EYE_TYPES, key: "eyeType"},
    {values: AVATAR_MOUTH_TYPES, key: "mouthType"},
  ],
  clothing: [
    {values: AVATAR_ACCESSORIES_TYPES, key: "accessoriesType", disabledOn: {topType: ["Eyepatch"]}},
    {values: AVATAR_CLOTHE_TYPES, key: "clotheType"},
    {values: AVATAR_CLOTHE_COLORS, key: "clotheColor", disabledOn: {clotheType: ["BlazerShirt", "BlazerSweater"]}},
    {
      values: AVATAR_GRAPHIC_TYPES,
      key: "graphicType",
      disabledOn: {clotheType: ["BlazerShirt", "BlazerSweater", "CollarSweater", "Hoodie", "Overall", "ShirtCrewNeck", "ShirtScoopNeck", "ShirtVNeck"]},
    },
  ],
};
