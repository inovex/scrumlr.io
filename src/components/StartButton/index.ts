import StartButton from './StartButton';
export default StartButton;
export * from './StartButton';
