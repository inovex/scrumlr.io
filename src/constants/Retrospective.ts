import { AddCardTheme } from '../components/AddCard';
import { IconNames } from '../components/Icon';

export type ColumnType = 'positive' | 'negative' | 'actions';
export const DEFAULT_RETRO_MODE = 'positiveNegative';

export interface Activity {
  icon: IconNames;
  description: string;
}

export interface Column {
  name: string;
  type: ColumnType;
}

export interface ColumnConfiguration extends Column {
  id: string;
  voting: {
    enabled: boolean;
    displayed: boolean;
  };
  focus: {
    enabled: boolean;
    column: string;
    align: 'left' | 'right';
  };
  sorted: boolean;
}

export interface PhaseConfiguration {
  name: string;
  description: string;
  columns: { [key: string]: ColumnConfiguration };
  activities: Activity[];
}

export function getTheme(type: ColumnType): AddCardTheme {
  switch (type) {
    case 'negative':
      return 'dark';
    case 'actions':
      return 'mint';
    case 'positive':
    default:
      return 'light';
  }
}

export function getInversedTheme(type: ColumnType): AddCardTheme {
  switch (type) {
    case 'negative':
      return 'light';
    case 'actions':
    case 'positive':
    default:
      return 'dark';
  }
}
