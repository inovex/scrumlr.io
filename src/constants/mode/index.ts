import leanCoffeeModeConfiguration from './LeanCoffeeMode';
import madSadGladConfiguration from './MadSadGladMode';
import kalmConfiguration from './KALM';
import simpleRetroModeConfiguration from './PositiveNegativeMode';
import startStopContinueRetroModeConfiguration from './StartStopContinueMode';
import { PhaseConfiguration } from '../Retrospective';

interface RetroModeConfiguration {
  lean: PhaseConfiguration[];
  positiveNegative: PhaseConfiguration[];
  startStopContinue: PhaseConfiguration[];
  madSadGlad: PhaseConfiguration[];
  kalm: PhaseConfiguration[];
}

export const retroModes: RetroModeConfiguration = {
  lean: leanCoffeeModeConfiguration,
  madSadGlad: madSadGladConfiguration,
  positiveNegative: simpleRetroModeConfiguration,
  startStopContinue: startStopContinueRetroModeConfiguration,
  kalm: kalmConfiguration
};

export type RetroMode = keyof RetroModeConfiguration;

export default retroModes;
