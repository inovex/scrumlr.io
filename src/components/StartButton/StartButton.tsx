import * as React from 'react';
import { RetroMode } from '../../constants/mode';
const reactDDMenu = require('react-dd-menu');
const DropdownMenu = reactDDMenu.DropdownMenu;

import './StartButton.scss';
import Icon from '../Icon';
import { DEFAULT_RETRO_MODE } from '../../constants/Retrospective';

export interface StartButtonProps {
  onStart: (retroMode: RetroMode) => void;
}

export interface StartButtonState {
  selectedMode: RetroMode;
  isSelectionMenuOpened: boolean;
}

const availableRetroModes = {
  lean: {
    label: 'Lean Coffee'
  },
  positiveNegative: {
    label: 'Simple Retro (Delta+)'
  },
  startStopContinue: {
    label: 'Start, Stop, Continue'
  },
  madSadGlad: {
    label: 'Mad, Sad, Glad'
  },
  kalm: {
    label: 'KALM (Keep, Add, Less, More)'
  }
};

export class StartButton extends React.Component<
  StartButtonProps,
  StartButtonState
> {
  constructor(props: StartButtonProps) {
    super(props);
    this.state = {
      selectedMode: DEFAULT_RETRO_MODE,
      isSelectionMenuOpened: false
    };
  }

  closeSelectionMenu = () => {
    this.setState({ isSelectionMenuOpened: false });
  };

  toggleSelectionMenu = () => {
    this.setState(state => ({
      isSelectionMenuOpened: !state.isSelectionMenuOpened
    }));
  };

  selectRetroMode = (retroMode: RetroMode) => () => {
    this.setState({ selectedMode: retroMode });
  };

  render() {
    const { onStart } = this.props;
    const { selectedMode, isSelectionMenuOpened } = this.state;

    const selectedModeConfiguration = availableRetroModes[selectedMode];

    const dropdownButton = (
      <button
        aria-label="Board mode selection dropdown"
        className="start-button__dropdown-toggle"
        onClick={this.toggleSelectionMenu}
      >
        <Icon name="more" />
      </button>
    );

    const ddMenuProps = {
      isOpen: isSelectionMenuOpened,
      close: this.closeSelectionMenu,
      toggle: dropdownButton,
      align: 'right',
      closeOnInsideClick: true
    };

    return (
      <div className="start-button">
        <button
          aria-label="Start session"
          className="start-button__start"
          onClick={() => onStart(selectedMode)}
        >
          <h1 className="start-button__start-label">Start</h1>
          <span className="start-button__start-mode">
            {selectedModeConfiguration.label}
          </span>
        </button>

        <DropdownMenu
          {...ddMenuProps}
          className="start-button__dropdown-wrapper"
        >
          {Object.keys(availableRetroModes).map((retroMode: RetroMode) => (
            <li key={retroMode}>
              <button
                onClick={this.selectRetroMode(retroMode)}
                aria-label={`Mode: ${availableRetroModes[retroMode].label}`}
              >
                {availableRetroModes[retroMode].label}
              </button>
            </li>
          ))}
        </DropdownMenu>
      </div>
    );
  }
}

export default StartButton;
