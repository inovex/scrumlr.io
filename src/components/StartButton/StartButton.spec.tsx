import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import { StartButton, StartButtonProps, StartButtonState } from './StartButton';
import { DEFAULT_RETRO_MODE } from '../../constants/Retrospective';
import { RetroMode } from '../../constants/mode';

describe('<StartButton />', () => {
  let wrapper: ShallowWrapper<StartButtonProps, StartButtonState>;
  let onStart: any;

  beforeEach(() => {
    onStart = jest.fn();
    wrapper = shallow(<StartButton onStart={onStart} />);
  });

  it('should match expected structure', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should callback with the default retro mode on initial click', () => {
    wrapper.find('.start-button__start').simulate('click');
    expect(onStart).toHaveBeenCalledWith(DEFAULT_RETRO_MODE);
  });

  it('should callback with the expected retro mode when it is changed', () => {
    const expectedRetroMode: RetroMode = 'lean';
    wrapper.setState({ selectedMode: expectedRetroMode });
    wrapper.find('.start-button__start').simulate('click');
    expect(onStart).toHaveBeenCalledWith(expectedRetroMode);
  });

  it('should have a closed selection menu initially', () => {
    expect(wrapper.state().isSelectionMenuOpened).toEqual(false);
  });
});
