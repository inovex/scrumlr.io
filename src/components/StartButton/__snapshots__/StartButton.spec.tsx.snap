// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StartButton /> should match expected structure 1`] = `
"<div className=\\"start-button\\">
  <button className=\\"start-button__start\\" onClick={[Function: onClick]}>
    <h1 className=\\"start-button__start-label\\">
      Start
    </h1>
    <span className=\\"start-button__start-mode\\">
      Simple Retro (Delta+)
    </span>
  </button>
  <DropdownMenu isOpen={false} close={[Function]} toggle={{...}} align=\\"right\\" closeOnInsideClick={true} className=\\"start-button__dropdown-wrapper\\" inverse={false} animAlign={{...}} textAlign={{...}} menuAlign={{...}} size={{...}} upwards={false} animate={true} enterTimeout={150} leaveTimeout={150} closeOnOutsideClick={true}>
    <li>
      <button onClick={[Function]}>
        Lean Coffee
      </button>
    </li>
    <li>
      <button onClick={[Function]}>
        Simple Retro (Delta+)
      </button>
    </li>
    <li>
      <button onClick={[Function]}>
        Start, Stop, Continue
      </button>
    </li>
    <li>
      <button onClick={[Function]}>
        Mad, Sad, Glad
      </button>
    </li>
    <li>
      <button onClick={[Function]}>
        KALM (Keep, Add, Less, More)
      </button>
    </li>
  </DropdownMenu>
</div>"
`;
